go/database/mpt/verification_proof.go

```diff
@@ -0,0 +1,336 @@
+	for _, addr := range correctAddresses {
+		if interrupt.IsCancelled(ctx) {
+			return interrupt.ErrCanceled
+		}
+
+		keys := generateUnknownKeys(numKeys)
+		proof, err := trie.CreateWitnessProof(addr, keys...)
+		if err != nil {
+			return err
+		}
+		if err := verifyStorageProof(rootHash, proof, addr, keys, nil); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// verifyAccountProof verifies the consistency between the input witness proofs and the account.
+func verifyAccountProof(root common.Hash, proof witness.Proof, addr common.Address, data accountData) error {
+	balance, complete, err := proof.GetBalance(root, addr)
+	if err != nil {
+		return err
+	}
+	if !complete {
+		return fmt.Errorf("proof incomplete for 0x%x", addr)
+	}
+	if got, want := balance, data.balance; got != want {
+		return fmt.Errorf("balance mismatch for 0x%x, got %v, want %v", addr, got, want)
+	}
+	nonce, complete, err := proof.GetNonce(root, addr)
+	if err != nil {
+		return err
+	}
+	if !complete {
+		return fmt.Errorf("proof incomplete for 0x%x", addr)
+	}
+	if got, want := nonce, data.nonce; got != want {
+		return fmt.Errorf("nonce mismatch for 0x%x, got %v, want %v", addr, got, want)
+	}
+	code, complete, err := proof.GetCodeHash(root, addr)
+	if err != nil {
+		return err
+	}
+	if !complete {
+		return fmt.Errorf("proof incomplete for 0x%x", addr)
+	}
+	if got, want := code, data.code; got != want {
+		return fmt.Errorf("code mismatch for 0x%x, got %v, want %v", addr, got, want)
+	}
+
+	return nil
+}
+
+// verifyStorageProof verifies the consistency between the input witness proofs and the storage.
+func verifyStorageProof(root common.Hash, proof witness.Proof, addr common.Address, keys []common.Key, storage map[common.Key]common.Value) error {
+	for _, key := range keys {
+		proofValue, complete, err := proof.GetState(root, addr, key)
+		if err != nil {
+			return err
+		}
+		if !complete {
+			return fmt.Errorf("proof incomplete for address: 0x%x, key: 0x%x", addr, key)
+		}
+		if got, want := proofValue, storage[key]; got != want {
+			return fmt.Errorf("storage mismatch for 0x%x, key 0x%x, got %v, want %v", addr, key, got, want)
+		}
+	}
+
+	return nil
+}
+
+// generateUnknownAddresses generates a slice of addresses that will never appear
+// in real MPT and the witness proof. Each address consists of four random
+// bytes, a number one, and the rest bytes are zero. The number one is set to
+// prevent generation of random address that has only zeros. Such an address
+// is reserved for a special contract.
+func generateUnknownAddresses(number int) []common.Address {
+	res := make([]common.Address, number)
+	for i := range res {
+		j := rand.Int()
+		res[i] = common.Address{byte(j), byte(j >> 8), byte(j >> 16), byte(j >> 24), 1}
+	}
+	return res
+}
+
+// generateUnknownKeys generates a slice of keys that will never appear in real
+// MPT and the witness proof. Each key consists of four random bytes, a number
```

Comment:
Same as for the addresses. They are unlikely, but possible. I would suggest to use them as random keys and compare the proof-content with the MPT content as also suggested for accounts.

Reply:
fixed by skipping potentially non-empty keys/addresses